2.x - Fix ApcEngine to work on PHP 7

jrbasso opened the pull request:
Currently Cake 2.8 is using only the `apc_*` functions, which doesn't work on PHP 7. The pecl `apcu` only support `apcu_*` functions on PHP 7+.

Since CakePHP 2.8 introduces the support to PHP 7, adding the support to both extensions.

The issue didn't show on CI before because apc wasn't being installed and the test case was skipping in this case.

jrbasso committed:
Supporting APCu on PHP 7

jippi commented:
can we just have two extensions? rather than a double method call to wrap it

in app.php you could just `engine = function_exists('apcu_dec') ? 'apcu' : 'apc'` and be much more efficient in number of calls

ADmad commented:
> can we just have two extensions?

The trade off would be lot of duplicate code.

dereuromark commented:
Having it in one seems a reasonable approach here for me. :+1:

ADmad commented:
Using something like below would avoid the `call_user_func_array()`  calls.

```php
$method = $this->_apcExtension . '_foo';  
$method(...);
```

jrbasso commented:
I agree that having 2 engines is overkill, specially when people migrate to Cake 3 which just have one.

@ADmad good point about the function by string. I did similar in one of the methods that required a param by reference. I will do it for all, it will be indeed faster and cleaner.

jrbasso committed:
Improved apc calls

Instead of calling the call_user_func_array it does the call directly. It is a little bit more verbose, but performs better.

jrbasso commented:
Change made.

markstory commented on .travis.yml:
```diff
@@ -41,6 +41,8 @@ before_script:
   - sh -c "if [ '$PHPCS' = '1' ]; then composer global require 'cakephp/cakephp-codesniffer:1.*'; fi"
   - sh -c "if [ '$PHPCS' = '1' ]; then ~/.composer/vendor/bin/phpcs --config-set installed_paths ~/.composer/vendor/cakephp/cakephp-codesniffer; fi"
   - echo "extension = memcached.so" >> ~/.phpenv/versions/$(phpenv version-name)/etc/php.ini
+  - if [[ ${TRAVIS_PHP_VERSION:0:3} == "7.0" ]] ; then print "yes" | pecl install apcu-5.1.3; else print "yes" | pecl install apcu-4.0.11; fi
```
Does apcu 5.x not work on PHP5?

jrbasso replied:
@markstory Nope. Will be really hard for any extension be compatible with 5 and 7. The internal PHP api changed a lot, they would have to define a lot of macros to make compatible, which makes the code really hard to maintain.

markstory commented:
Looks good to me.

Marlinc commented:
Why not create a method that can handle the two extensions? Something like 'public function callExtention($method)``` that would internally chose the right extension and get the parameters using ```func_get_args()```.

Its not a huge change but it might the code easier to read.

ADmad commented:
@Marlinc That's exactly what done initially but was changed to avoid extra method calls and using `call_user_func_array()`.

Marlinc commented:
@ADmad I'm sorry, haven't seen that.

ADmad commented:
No problem :)

jippi commented:
quite a `$func`y fix :P
